### PR TITLE
Enhance check-i18n script to minimize false positives and improve detection accuracy

### LINKED ISSUE #5051: Bug: `check-i18n`  false positives on TS types, CSS classes, and ignore not working
I am working on a PR and the `check-i18n` script blocked me with these errors:

```bash
src/screens/OrgList/OrgList.tsx:226 -> "): Promise"
src/screens/OrgList/OrgList.tsx:412 -> "${styles.orgImgContainer} shimmer"
src/screens/OrgList/OrgList.tsx:422 -> "shimmer ${styles.button}"
src/screens/OrgList/OrgList.tsx:504 -> "btn  btn-primary ${styles.pluginStoreBtn}"
src/screens/OrgList/OrgList.tsx:507 -> "orgstore/id=${dialogRedirectOrgId}"
```

These are false positives:

- Promise is a TypeScript return type, not text.
- shimmer and btn are CSS class names inside template literals.
- The last one is a URL path, not UI text.
- I tried to suppress these errors using `// i18n-ignore-line`, but the script ignores the comment and still fails.

I also noticed that if I put text on a new line, the script does not catch it.
Example:
```javascript
<div>
  This text is ignored by the script
</div>
```

### LINKED ISSUE #5062: Enhance i18n detection script to reduce false positives
**Description Problem**
PR [#4956](https://github.com/PalisadoesFoundation/talawa-admin/pull/4956) successfully implements the core i18n detection infrastructure (`scripts/check-i18n.js`), which identifies hardcoded user-visible text in the codebase. However, during testing on the production codebase, we've identified false positives that will block legitimate development work when the pre-commit hook is active.

**Common False Positive Patterns (few examples)**
The current script incorrectly flags (there can be other incorrect flags as well) the following as violations:

1. **TypeScript type annotations**
   ```tsx
   const handler = (): string => { ... }  // Flags "string" as violation
   const fn = (): Promise<void> => { ... }  // Flags "void" as violation
   ```

2. **CSS utility classes in `className` attributes**
   ```tsx
   <div className={`btn primary m-3 p-2`} />  // Flags "btn", "primary"
   <i className="fi fi-rr-home" />  // Flags font icon classes
   ```

3. **Technical ARIA attributes**
   ```tsx
   <div aria-hidden="true" role="button" data-testid="my-test" />
   ```

4. **Date format strings**
   ```tsx
   <input placeholder="YYYY-MM-DD" />  // Technical format, not user text
   ```

5. **Regular expression patterns**
   ```tsx
   const pattern = "[a-z]+";  // Technical pattern, not user text
   ```

6. **Developer-facing messages**
   ```tsx
   console.log("Debug info");  // Developer message, not user-facing
   throw new Error("Internal error");  // Error message, not UI text
   ```

7. **GraphQL queries, JSON operations, string methods**
   ```tsx
   const query = gql`query { ... }`;
   JSON.stringify({ key: "value" });
   ```

8. **JavaScript comparison operators in JSX**
   ```tsx
   {age >= 18 && age <= 40}  // Logical comparison, not text
   ```

## Related
- Prerequisite PR (i18n detection infrastructure) https://github.com/PalisadoesFoundation/talawa-admin/pull/4956

### PR DESCRIPTION
**Pull request by rawadhossain**
**What kind of change does this PR introduce?**
`bugfix`- This PR enhances the `check-i18n` script to significantly reduce false positives while preserving accurate detection of user-visible text. The improvements were driven by failures observed in code and reported in #5051

**Issue Number:**
Fixes #5062
Fixes #5051 

Related PR https://github.com/PalisadoesFoundation/talawa-admin/pull/4956

**Summary**
Brief summary on the PR:
- **Multi-line JSX text detection**: Detects user-visible text spanning multiple lines between JSX tags.
- **Improved template literal handling**: Detects individual template literals inside JSX expressions instead of treating the entire expression as text.
- **Nested template literal support**: Correctly handles nested template literals (e.g. `` `outer ${`inner`} text` ``) without partial or incorrect matches.
- Split oversized test file to comply with CI limits and improve maintainability:
  - `check-i18n-basic.test.js` – core detection behavior
  - `check-i18n-enhanced.test.js` – enhanced and edge case coverage
  - `check-i18n.test-utils.js` – shared test utilities
- **Updated `false-positives.tsx`** to include only legitimate false-positive cases.
- **Removed unused `mixed.tsx` fixture**, as tests now generate inline content.
- Fixed incorrect partial matches in nested template literals.
- Corrected template literal regex to detect individual literals rather than full JSX expressions.

**Additional Fix**
Fixed a flaky test in `ActionItemModal.spec.tsx` that was failing in CI due to timing issues. Added proper async waits for form initialization.

**Does this PR introduce a breaking change?**
No 
### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

## Summary by CodeRabbit

* **Improvements**
  * Enhanced i18n checker with broader context-aware detection, smarter skipping of technical patterns (URLs, dates, regex, attributes, class names) and improved CLI output/word counting.

* **New Fixtures**
  * Added a comprehensive false-positives fixture to validate ignored patterns.
  * Removed a mixed-content fixture.

* **Tests**
  * Expanded and reorganized test suites with many edge cases and dynamic temp-file testing; updated an async test to wait for populated input.

* **Chores**
  * Introduced shared test utilities for running the checker and managing temp files.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: scripts/check-i18n.js [hunk 1]
```diff
@@ -66,13 +141,136 @@ const stripComments = (content) =>
...
+  // Check previous line (for i18n-ignore-next-line)
+  if (lineIndex > 0) {
+    const prevLine = originalLines[lineIndex - 1];
+    if (/\/\/\s*i18n-ignore-next-line/i.test(prevLine)) {
+      return true;
+    }
+  }
+
+  return false;
+};
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply scripts/check-i18n.js]: The ignore mechanism is intentionally limited to //-style comments (i18n-ignore-line / i18n-ignore-next-line). JSX comment syntax ({/* … */}) is not currently supported by design, as the script operates on line-based heuristics rather than JSX AST parsing.

File: scripts/check-i18n.js [hunk 2]
```diff
@@ -66,13 +141,136 @@ const stripComments = (content) =>
...
+
+  // Skip GraphQL queries (gql`...`)
+  if (/gql\s*`/.test(beforeMatch)) {
+    return true;
+  }
+
+  // Skip new RegExp(...) or /pattern/ regex literals
+  if (/new\s+RegExp\s*\(/.test(beforeMatch) || /\/[^\/]+\//.test(beforeMatch)) {
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply scripts/check-i18n.js:345]: This check is intentionally permissive and operates as a heuristic guard inside isInSkipContext, not as a strict regex literal detector. JSX text such as Go to /admin/settings is still evaluated correctly through later JSX text handling and URL heuristics, and does not get skipped incorrectly. I’d prefer to keep the current balance.

File: scripts/check-i18n.js [hunk 3]
```diff
@@ -97,32 +314,255 @@ const collectViolations = (filePath) => {
...
+          } else {
+            // Closing backtick
+            backtickDepth--;
+            if (backtickDepth === 0) {
+              // Complete template literal found
+              const fullText = line.substring(templateStart + 1, i);
+              templateLiterals.push({ fullText, matchIndex: templateStart });
+              inTemplateLiteral = false;
+            }
+          }
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply scripts/check-i18n.js:632]: The current check intentionally matches the actual JavaScript grammar for nested template literals, where the ${ token is not whitespace-tolerant. Whitespace between $ and { is not valid syntax, and whitespace inside the ${ ... } expression does not affect detection of nested backticks.

**Comment by rawadhossain** [comment]
> move the tests to a folder under src/test-utils where they are better placed

Noted. Moving the files in next commit.

**Comment by palisadoes** [comment]
@coderabbitai full review and reevaluate all resolved items and include the following points:

1. **Test Coverage & Quality Validation:**
   1. Validate whether any updated test files:
      1. Cover all use cases including edge cases, error scenarios, and success paths
      2. Have valid, properly structured tests following Jest and React Testing Library best practices
      3. Provide 100% test code coverage for all modified/added functions, components, and branches
      4. Include proper mocking of GraphQL queries/mutations, external dependencies, and API calls
      5. Test both user interactions and state changes comprehensively
      6. Require any additional tests for uncovered scenarios
   2. For each modified file with logic or UI, verify corresponding test files exist and are updated
   3. Check that test files follow naming conventions (*.test.tsx, *.test.ts, *.spec.tsx, *.spec.ts)
   4. Validate that tests don't include disabled tests (it.skip, describe.skip) unless explicitly documented
   5. Ensure mock data cleanup is performed (as required by pre-commit hooks)

2. **Code Quality & Repository Standards:**
   1. Verify TypeScript compliance:
      1. No use of `any` type without explicit justification
      2. Proper interface/type definitions for all props, state, and function parameters
      3. No TypeScript errors or warnings
   2. Validate GraphQL implementation:
      1. Queries/mutations are properly typed and follow existing patterns in src/GraphQl/
      2. No duplicate or redundant queries (check for pagination inconsistencies)
      3. Error handling is implemented for all GraphQL operations
   3. Check React component best practices:
      1. Functional components with proper hooks usage
      2. Material-UI component patterns are followed consistently
      3. No prop drilling; proper use of context when needed
      4. Proper cleanup in useEffect hooks
   4. Verify internationalization (i18n):
      1. All user-facing strings use i18next translation keys
      2. New translation keys are added to locale files
      3. No hardcoded English text in UI components
   5. Validate accessibility:
      1. Proper ARIA labels and roles
      2. Keyboard navigation support
      3. Screen reader compatibility

3. **Documentation & Code Comments:**
   1. Verify auto-generated documentation is updated (as enforced by Check-AutoDocs job)
   2. Check for adequate JSDoc comments on public functions and complex logic
   3. Ensure README.md or related docs are updated if new features are added
   4. Validate that Table of Contents is updated (as required by pre-commit hooks)

4. **Identify Issues:**
   1. Lines of code that are missing tests (list specific line numbers)
   2. Any unnecessary files that have been submitted (build artifacts, node_modules, .env files, IDE configs)
   3. Unused imports or dead code (as checked by knip)
   4. Console.log statements or debug code left in production files
   5. Files that should be excluded from the PR scope

5. **PR Requirements Validation:**
   1. Verify the PR meets the requirements of the linked issue:
      1. All acceptance criteria are addressed
      2. No out-of-scope changes are included
   2. Validate compliance with all previously suggested improvements and comments:
      1. All review comments have been addressed
      2. Requested changes have been implemented correctly
   3. Check PR metadata:
      1. Proper title and description
      2. Linked to appropriate issue(s)
      3. Target branch is correct (typically `develop`)
   4. Verify no sensitive information is exposed (as checked by Check-Sensitive-Files job)

6. **CI/CD Pipeline Compatibility:**
   1. Validate that code changes won't break any of the 17+ CI jobs
   2. Ensure pre-commit hook requirements are met:
      1. Code formatting (Prettier)
      2. Linting (ESLint)
      3. Type checking (TypeScript)
   3. Check that changes don't introduce performance regressions for the 12-shard test execution

7. **Security & Best Practices:**
   1. No security vulnerabilities introduced
   2. No exposed API keys, tokens, or credentials
   3. Proper input validation and sanitization
   4. No SQL injection or XSS vulnerabilities in any data handling

8. **Review Decision:**
   - If the submitted code requires **any changes** (even trivial ones such as missing tests, typing issues, documentation gaps, or minor improvements):
     1. Provide detailed, actionable feedback for each issue
     2. **Explicitly change the PR status to CHANGES_REQUESTED using:** `gh pr review <PR_NUMBER> --request-changes -b "<summary of required changes>"`
   - If **no changes are required** and all criteria above are satisfied:
     1. **Explicitly approve the PR using:** `gh pr review <PR_NUMBER> --approve -b "All validation checks passed. LGTM! ✅"`

9. **Focus Areas:**
   - Prioritize substantive issues over minor stylistic suggestions
   - Focus on issues that would cause CI pipeline failures or production bugs
   - Highlight any deviations from established codebase patterns

**Review by palisadoes — CHANGES_REQUESTED** [review]
Please address this:
- https://github.com/PalisadoesFoundation/talawa-admin/pull/5066#issuecomment-3662252578

**Comment by palisadoes** [comment]
PTAL. This explains the importance of our pre-commit hooks and why bypassing them can lead to a frustrating PR experience.

- https://docs-admin.talawa.io/docs/developer-resources/testing/#pre-commit-hooks

File: scripts/check-i18n.js [hunk 4]
```diff
@@ -97,32 +487,255 @@ const collectViolations = (filePath) => {
...
+      if (
+        /(>=|<=|==|!=|===|!==|&&|\|\|)\s*\d+/.test(text) || // Comparison with numbers
+        /\d+\s*(>=|<=|==|!=|===|!==|&&|\|\|)/.test(text) || // Number before operator
+        /(return|const|let|var|if|while|for)\s+.*(>=|<=|==|!=|===|!==)/.test(
+          beforeMatch,
+        ) || // Code keywords before comparison
+        /\.(filter|map|reduce|find|some|every)\s*\(/.test(beforeMatch) // Array methods
+      ) {
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply scripts/check-i18n.js:548]: I agree that some of the heuristics overlap conceptually, but in this PR they are intentionally kept inline to preserve local clarity and avoid introducing additional abstraction while fixing detection issues. Since the duplicated logic serves slightly different contextual purposes, extracting helpers here would expand the scope and risk unintended behavior changes.

**Comment by rawadhossain** [comment]
yeah, you are right. I found there is a flaky test in `ActionItemModal.spec.tsx` (unrelated to this PR) that pass locally but fail in CI due to timing issues (async operations). Fixing it by adding proper async wait for form initialization.

**Comment by palisadoes** [comment]
@coderabbitai full review using this prompt:
- https://github.com/PalisadoesFoundation/talawa-admin/pull/5066#issuecomment-3662237046

**Review by palisadoes — APPROVED** [review]
